Restructure extensions

kynikos opened the pull request:
This PR restructures extensions and better integrates them into the rest of the code. There are also some other related changes.

Since the way in which extensions are now installed and configured is much different from before, I thought it would be better to inform the users instead of trying to preserve backwards compatibility, however I can add some conversion mechanism if you prefer.

Before the configuration was something like:
```
keys = [Key(["mod4"], 'm', lazy.run_extension(extension.DmenuRun))]
extensions = {
    'dmenu': {...}
}
```
Now extensions share the default configuration with widgets, since integrating and improving the consistency of their appearance is what this is all about I think. I thought I'd rename `widget_defaults` to `extension_defaults` because the name is more generic, and the bar could be restructured as a normal extension in the future, which would be a step forward towards implementing #622. Being a historical option, `widget_defaults` is still supported as an alias to `extension_defaults`. Extensions should now be installed like this:
```
keys = [Key(["mod4"], 'm', lazy.run_extension(extension.DmenuRun(...)))]
extension_defaults = {
    ...
}
```

I've tried to add descriptions to the commits that needed some explanation, please ask questions if something's not clear.

kynikos commented:
I've added two commits (the others are unchanged, in case somebody had started reviewing them):
* aed7693 improves the behavior of the `lines` option when used by the `WindowList` extension
* 95a01b4 amends 3214079 restoring full backwards compatibility, to hopefully facilitate the merge of this branch

flacjacket commented on libqtile/confreader.py:
```diff
@@ -87,17 +87,39 @@ def __init__(self, fname=None, is_restart=False):
+        # Keep supporting the deprecated 'widget_defaults' option
+        # TODO: Perhaps completely deprecate it in the future
+        if hasattr(config, "widget_defaults"):
+            deprecated_defaults.update(config.widget_defaults)
```
When we hit these, we should throw a deprecation warning into the log

kynikos replied:
I've added c97eab1

flacjacket commented on CHANGELOG:
```diff
@@ -1,6 +1,9 @@
-    !!! extention subpackage renamed to extension !!!
-    !!! extentions option renamed to extensions !!!
+    !!! `extention` subpackage renamed to `extension` !!!
+    !!! `widget_defaults` and `extentions` options merged into `extension_defaults` !!!
```
I see why it would be nice to unify these, but it might be better to pick a name that is neither specific to widgets or extensions. Particularly because widget_defaults will be the most prevalent for users.

kynikos replied:
Sure, I can think of `option_defaults`, `variable_defaults`, `arg_defaults`, `par_defaults`, simply `defaults`, and all the reversed `default_*` variants, e.g. `default_options` etc.

kynikos replied:
...however the idea to use `extension_defaults` was based on the possibility to restructure the bar as an extension on day. More ideas, if we want to agree to call the extension+bar group as "plugins" or "add-ons", the variable could be called `plugin_defaults` or `addon_defaults`.

zordsdavini commented:
should we have one package or two after split? For example, we could have
qtile-bar with its widgets and qtile-plugins

On 24 March 2017 at 13:29, Dario Giovannetti <notifications@github.com>
wrote:

> *@kynikos* commented on this pull request.
> ------------------------------
>
> In CHANGELOG
> <https://github.com/qtile/qtile/pull/1026#discussion_r107883028>:
>
> > @@ -1,6 +1,9 @@
>  qtile next_version:
> -    !!! extention subpackage renamed to extension !!!
> -    !!! extentions option renamed to extensions !!!
> +    !!! `extention` subpackage renamed to `extension` !!!
> +    !!! `widget_defaults` and `extentions` options merged into `extension_defaults` !!!
>
> ...however the idea to use extension_defaults was based on the
> possibility to restructure the bar as an extension on day. More ideas, if
> we want to agree to call the extension+bar group as "plugins" or "add-ons",
> the variable could be called plugin_defaults or addon_defaults.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qtile/qtile/pull/1026#discussion_r107883028>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AD_1PIDFO-DUgATaTNFGRZKIeegQ5r5Qks5ro6kFgaJpZM4MhuKS>
> .
>



-- 
Udovīčė Arns

kynikos commented:
@zordsdavini let's discuss that in #622, none of the repo admins have ever commented about that, it's only my idea

kynikos commented:
Question: I created 1 new file in this PR, `libqtile/extension/base.py`, and I put Aldo Cortesi in the copyright line at the top, however I've noticed that it's more common to add the file creator's name, so can/should I change it?

kynikos commented:
Since there were conflicts with the recently merged commits, I've taken the chance to rebase this PR squashing some fixup commits. For reference, the original commits are in https://github.com/qtile/qtile/compare/develop...kynikos:extensions-old

More importantly, I've renamed `extension_defaults` to a more neutral `addon_defaults`.

kynikos commented:
Also, 9b861de is designed so it can be left unmerged if you prefer, i.e. keeping `widget_defaults` and `extension_defaults` separate.

This makes me think of yet another option: in `default_config.py`, `extension_defaults` could be defined after `widget_defaults` as:
```
widget_defaults = dict(
    ...
)
extension_defaults = widget_defaults
```
or:
```
widget_defaults = dict(
    ...
)
extension_defaults = widget_defaults.copy()
```
And let users decide what to do with them.

zordsdavini commented:
Extensions are new thing, so I think we can drop support for fallback. If we move everything to option_defaults, then it would be needful to fallback widget_defaults as you showed above

zordsdavini commented:
with this PR we can have 0.11 version of Qtile

kynikos commented:
I've rebased this on the new merged commits, is there anything I can do to help merging this patch?

zordsdavini commented on libqtile/extension/dmenu.py:
```diff
@@ -19,106 +19,149 @@
+                selected_foreground="#fff",
+                height=24,  # Only supported by some dmenu forks
+            ))),
         ]
```
It looks like if you'll have more running extensions with dmenu, then you'll need to duplicate configuration. That's why I coded extension_defaults. To have same look of dmenu everywhere. What do you think about single place of dmenu configuration?

kynikos replied:
I would leave specific configuration grouping to the users: considering that `config.py` is a plain Python script it will be possible to create for example:
```
dmenu_config = {
    ...
}
extension.DmenuRun(foo='foo', **dmenu_config)
extension.WindowList(bar='bar', **dmenu_config)
```
Dictionaries can also be merged with `update()` etc., so there's plenty of possibilities to avoid duplicating values according to one's own needs. My opinion is that this is Qtile's configuration file, not dmenu's or some other application's, and I'd like to keep it simple by default.

What I'd be more open to would be having option prefixes, e.g. `dmenu-height` or `dmenu-selected-foreground`.

kynikos commented:
So, with the hope of making it easier to merge the patch, I've changed it to leave `widget_defaults` untouched, and now there's a separate `extension_defaults` configuration variable which by default copies `widget_defaults`. In short, nothing at all changes anymore for users who don't use extensions!

I've also added `dmenu_` prefixes to some dmenu-specific options, to make it easier to configure parameters only for dmenu-based extensions. The `font` option now acts like the one for widgets, and I've added an analogous `fontsize` option, and a `dmenu_font` option to optionally override them using dmenu's `-fn` syntax.

I've also taken the chance to add e98c403 which replaces the remaining instances of Arial as a default font (weird on a program meant to run on Linux).

flacjacket commented:
Thanks for continuing to push on this for almost one month, this looks great, thanks for the hard work!

kynikos commented:
Cheers! However I have to apologize because I stopped properly testing the WindowList extension at a certain point, and I've just realized that the merged patch breaks it... kynikos@27bc20f fixes it, can you please merge that too, or do you need a new PR? (extensions are not pytested, so that commit isn't going to break the tests)

flacjacket commented:
Got it! Thanks for the fix!
